### PR TITLE
Feat/relations in es

### PR DESCRIPTION
**Pull request by suraj5077**
## Relationship events on Kafka enabled
Notification events on ATLAS_ENTITIES topic 

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [x] Lint rules pass locally
- [x] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasEncodingUtil.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package org.apache.atlas.repository.graphdb.janus;
+
+import org.janusgraph.util.encoding.LongEncoding;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Objects;
+
+public class AtlasEncodingUtil {
```
Comment (n5nk) [inline graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasEncodingUtil.java:9]: - Move the class under `utils` package
- Rename the class to `AtlasUtils` or simply `Utils` with method name as `toLongEncoding()`
- nit: Add `@javax.annotation.Nonnull` annotation in the params `toLongEncoding(@Nonnull Object obj)`.

File: .github/workflows/codeql.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # For most projects, this workflow file will not need changing; you simply need
-# to commit it to your repository. 
+# to commit it to your repository.
```
Comment (n5nk) [inline .github/workflows/codeql.yml:2]: revert the file

File: .github/workflows/codeql.yml [hunk 3]
```diff
@@ -11,7 +11,7 @@
 #
 name: "CodeQL"
 
-on: 
+on:
```
Comment (n5nk) [inline .github/workflows/codeql.yml:14]: revert the file

File: intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java [hunk 4]
```diff
@@ -85,11 +88,13 @@ public class AtlasRelationship extends AtlasStruct implements Serializable {
     private Date          updateTime     = null;
     private Long          version        = 0L;
 
-    public enum Status { ACTIVE, DELETED }
+    public enum Status { ACTIVE, DELETED, PURGE_DUE_TO_ENTITY }
```
Comment (n5nk) [inline intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java:91]: Can you add details or comments in the code about what is `PURGE_DUE_TO_ENTITY`

File: intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java [hunk 5]
```diff
@@ -85,11 +88,13 @@ public class AtlasRelationship extends AtlasStruct implements Serializable {
...
 
     private Set<AtlasClassification> propagatedClassifications;
     private Set<AtlasClassification> blockedPropagatedClassifications;
 
+    private Map<String, Object> customRelationshipInfo;
```
Comment (n5nk) [inline intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java:96]: I think the variable `customRelationshipInfo` isnt used.

File: intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationshipHeader.java [hunk 6]
```diff
@@ -80,6 +84,10 @@ public AtlasRelationshipHeader(AtlasRelationship relationship) {
             case DELETED:
                 setStatus(AtlasEntity.Status.DELETED);
                 break;
+
+            case PURGE_DUE_TO_ENTITY:
```
Comment (n5nk) [inline intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationshipHeader.java:88]: can you rename to be more specific `PURGE_DUE_TO_ENTITY` seems incomplete.

File: repository/pom.xml [hunk 7]
```diff
@@ -309,6 +309,12 @@
             <artifactId>java-statsd-client</artifactId>
             <version>3.1.0</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.atlas</groupId>
+            <artifactId>atlas-graphdb-janus</artifactId>
+            <version>3.0.0-SNAPSHOT</version>
```
Comment (n5nk) [inline repository/pom.xml:315]: nit: Move version to `<properties>` section

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 8]
```diff
@@ -998,15 +998,24 @@ protected void deleteVertex(AtlasVertex instanceVertex, boolean force) throws At
 
         // Delete external references to this vertex - incoming edges from lineage or glossary term edges
         final Iterable<AtlasEdge> incomingEdges    = instanceVertex.getEdges(AtlasEdgeDirection.IN);
+        final Iterable<AtlasEdge> outgoingEdges    = instanceVertex.getEdges(AtlasEdgeDirection.OUT);
         final boolean             isPurgeRequested = RequestContext.get().isPurgeRequested();
 
+        if (!RequestContext.get().getDeleteType().equals(DeleteType.SOFT)) {
```
Comment (n5nk) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1004]: nit: We can simply add `isSoftDelete` or `isHardDelete` or `isPurge` etc with in `DeleteType` enum to void boiler plate code and re-write the code as `RequestContext.get().getDeleteType().isSoftDelete()`

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 9]
```diff
@@ -998,15 +998,24 @@ protected void deleteVertex(AtlasVertex instanceVertex, boolean force) throws At
...
+            for (AtlasEdge edge : outgoingEdges) {
+                if (isRelationshipEdge(edge))
+                    AtlasRelationshipStoreV2.saveRelationshipDeletionContext(RequestContext.get().getDeleteType(), null, edge, edge.getOutVertex(), edge.getInVertex(), entityRetriever);
```
Comment (n5nk) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1007]: Lets not pass `null` as arguments, rather overload the method `saveRelationshipDeletionContext`

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java [hunk 10]
```diff
@@ -998,15 +998,24 @@ protected void deleteVertex(AtlasVertex instanceVertex, boolean force) throws At
...
+        }
+
         for (AtlasEdge edge : incomingEdges) {
             AtlasEntity.Status edgeStatus = getStatus(edge);
             boolean            isProceed   = edgeStatus == (isPurgeRequested ? DELETED : ACTIVE);
 
             if (isProceed) {
                 if (isRelationshipEdge(edge)) {
                     deleteRelationship(edge);
+                    AtlasRelationshipStoreV2.saveRelationshipDeletionContext(RequestContext.get().getDeleteType(), null, edge, edge.getOutVertex(), edge.getInVertex(), entityRetriever);
```
Comment (n5nk) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1018]: same as above. Overload `saveRelationshipDeletionContext`

File: intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java (hunk 5 above)
Comment (suraj5077) [reply intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java:96]: It is used in the Kafka message. Check setCustomRelationshipInfo(customRelationshipInfo) in AtlasRelationshipHeader method.

File: .github/workflows/codeql.yml (hunk 2 above)
Comment (suraj5077) [reply .github/workflows/codeql.yml:2]: Done

File: .github/workflows/codeql.yml (hunk 3 above)
Comment (suraj5077) [reply .github/workflows/codeql.yml:14]: Done

File: graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasEncodingUtil.java (hunk 1 above)
Comment (suraj5077) [reply graphdb/janus/src/main/java/org/apache/atlas/repository/graphdb/janus/AtlasEncodingUtil.java:9]: Cannot move to utils package. This class has JanusGraph dependency.

File: intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java (hunk 4 above)
Comment (suraj5077) [reply intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationship.java:91]: Done

File: intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationshipHeader.java (hunk 6 above)
Comment (suraj5077) [reply intg/src/main/java/org/apache/atlas/model/instance/AtlasRelationshipHeader.java:88]: Done

File: repository/pom.xml (hunk 7 above)
Comment (suraj5077) [reply repository/pom.xml:315]: added at project level now

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java (hunk 9 above)
Comment (suraj5077) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v1/DeleteHandlerV1.java:1007]: Done

**Pull request merged**